buildSrc/subprojects/binary-compatibility/src/main/kotlin/org/gradle/binarycompatibility/metadata/IsKotlinInternalMetadataQuery.kt

```diff
@@ -0,0 +1,305 @@
+    } as T
+
+
+private
+interface IsKotlinInternal {
+    var isKotlinInternal: Boolean
+}
+
+
+private
+class IsInternalTypeKmClassVisitor(
+    private val jvmSignature: String
+) : KmClassVisitor(), IsKotlinInternal {
+
+    override var isKotlinInternal = false
+
+    private
+    var isDoneVisiting = false
+
+    override fun visit(flags: Flags, name: ClassName) {
+        if (!isDoneVisiting && jvmSignature == name.replace("/", ".")) {
+            isKotlinInternal = flags.isInternal
+            isDoneVisiting = true
+        }
+    }
+}
+
+
+private
+class IsInternalFieldKmClassVisitor(
+    private val jvmSignature: String
+) : KmClassVisitor(), IsKotlinInternal {
+
+    override var isKotlinInternal = false
+
+    private
+    var isDoneVisiting = false
+
+    private
+    val onMatch = { isInternal: Boolean ->
+        isKotlinInternal = isInternal
+        isDoneVisiting = true
+    }
+
+    override fun visitProperty(flags: Flags, name: String, getterFlags: Flags, setterFlags: Flags): KmPropertyVisitor? =
+        if (isDoneVisiting) null
+        else IsInternalMemberKmPropertyExtensionVisitor(jvmSignature, flags, getterFlags, setterFlags, onMatch)
+
+    override fun visitExtensions(type: KmExtensionType): KmClassExtensionVisitor? =
+        if (isDoneVisiting) null
+        else object : JvmClassExtensionVisitor() {
+            override fun visitLocalDelegatedProperty(flags: Flags, name: String, getterFlags: Flags, setterFlags: Flags): KmPropertyVisitor? =
+                IsInternalMemberKmPropertyExtensionVisitor(jvmSignature, flags, getterFlags, setterFlags, onMatch)
+        }
+}
+
+
+private
+class IsInternalConstructorKmClassVisitor(
+    private val jvmSignature: String
+) : KmClassVisitor(), IsKotlinInternal {
+
+    override var isKotlinInternal = false
+
+    private
+    var isDoneVisiting = false
+
+    override fun visitConstructor(flags: Flags): KmConstructorVisitor? =
+        if (isDoneVisiting) null
+        else object : KmConstructorVisitor() {
+            override fun visitExtensions(type: KmExtensionType): KmConstructorExtensionVisitor =
+                object : JvmConstructorExtensionVisitor() {
+                    override fun visit(desc: JvmMethodSignature?) {
+                        if (jvmSignature == desc?.asString()) {
+                            isKotlinInternal = flags.isInternal
+                            isDoneVisiting = true
+                        }
+                    }
+                }
+        }
+}
+
+
+private
+class IsInternalMethodKmClassVisitor(
+    private val jvmSignature: String
+) : KmClassVisitor(), IsKotlinInternal {
+
+    override var isKotlinInternal = false
+
+    private
+    var isDoneVisiting = false
+
+    private
+    val onMatch = { isInternal: Boolean ->
+        isKotlinInternal = isInternal
+        isDoneVisiting = true
+    }
+
+    override fun visitFunction(flags: Flags, name: String): KmFunctionVisitor? =
+        if (isDoneVisiting) null
```

Comment:
⭕️ Code could also short-circuit on `!flags.isInternal`, right? As in `if (isDoneVisiting || !flags.isInternal) null`

Reply:
Yep it could, but not uniformly over the visitor api, I left it like that for clarity